#' Call single nucleotide polymorphism variants
#'
#' This function calls SNP variants from BAM files. The results are formatted
#' into a VCF file and the ID column is populated with dbSNP IDs.
#'
#' @param bam A list of paths to BAM files
#' @param threads The number of cores to use.
#' @param outputdir The output directory for the VCF file.
#' @return A list of file paths to the VCF files.
#' @export
callindels <- function(bam,
                       threads = 4L,
                       outputdir = getwd()) {
  if (Sys.info()[["sysname"]] != "Linux") {
    stop("callindels() only works on Linux! Please run it in a Linux environment.")
  }
  cat(paste0("These are the species currently supported by exvar: \n",
             "[1] Homo sapiens (hg19) \n",
             "[2] Homo sapiens (hg38) \n",
             "[3] Mus musculus \n",
             "[4] Arabidopsis thaliana \n",
             "[5] Drosophila melanogaster \n",
             "[6] Danio rerio \n",
             "[7] Rattus norvegicus \n",
             "[8] Saccharomyces cerevisiae \n",
             "[9] Caenorhabditis elegans \n"))
  species <- readline("Type the number of the species that you would like to use as a reference: ")
  cat("Enable single-chromosome splitting? \nNote: enabling splitting reduces the RAM requirements, but may take longer.")
  mode <- readline("Type [y/n] for [yes/no]: ")
  wd <- getwd()

  if (mode == "y") {
    ##Sets the reference genome that corresponds to the species chosen by the user
    switch(species,
           "1"={
             #library(BSgenome.Hsapiens.UCSC.hg19)
             #library(XtraSNPlocs.Hsapiens.dbSNP144.GRCh37)
             ##Homo sapiens hg19
             if (!requireNamespace("BSgenome.Hsapiens.UCSC.hg19", quietly = TRUE)) {
               stop("The package 'BSgenome.Hsapiens.UCSC.hg19' is required but not installed.
       Install it with: BiocManager::install('BSgenome.Hsapiens.UCSC.hg19')")
             }
             if (!requireNamespace("XtraSNPlocs.Hsapiens.dbSNP144.GRCh37", quietly = TRUE)) {
               stop("The package 'XtraSNPlocs.Hsapiens.dbSNP144.GRCh37' is required but not installed.
       Install it with: BiocManager::install('XtraSNPlocs.Hsapiens.dbSNP144.GRCh37')")
             }
             organism <- BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19

             ##Selects hg19 as the reference genome
             ##If reference doesn't exist within package directory, create one
             if (dir.exists(paste0(find.package("exvar"), "/hg19"))) {
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"))
             } else {
               print("Reference genome not found. Creating reference. This might take a while...")
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"),
                                    create = TRUE)
             }
             output <- c()
             for (i in bam) {
               print(paste0("Analysing ", basename(i)))
               if (isFALSE(file.exists(paste0(i, ".bai")))) {
                 print("Indexing bam...")
                 setwd(dirname(i))
                 Rsamtools::sortBam(i, tools::file_path_sans_ext(i))
                 Rsamtools::indexBam(i)
               }
               setwd(wd)
               bamfl <- Rsamtools::BamFile(i)
               chromosomes <- GenomeInfoDb::seqlevels(bamfl)
               print("Splitting BAM into chromosomes...")
               tempfolder <- paste0(dirname(i), "/temp")
               dir.create(tempfolder)
               for (x in chromosomes) {
                 print(paste0("Creating ",
                              basename(tools::file_path_sans_ext(i)), "_", x, ".bam ",
                              "in temp folder."))
                 system(paste0("samtools view -b ",
                               i,
                               " ", x, " > ",
                               tempfolder, "/",
                               basename(tools::file_path_sans_ext(i)), "_", x, ".bam"))
                 setwd(tempfolder)
                 Rsamtools::sortBam(paste0(basename(tools::file_path_sans_ext(i)), "_", x, ".bam"),
                         paste0(basename(tools::file_path_sans_ext(i)), "_", x))
                 Rsamtools::indexBam(paste0(basename(tools::file_path_sans_ext(i)), "_", x, ".bam"))
                 setwd(wd)
               }
               chrbam <- tools::list_files_with_exts(tempfolder, "bam")
               chrbam <- sort(chrbam)


               for (bam in chrbam) {
                 print(paste0("Identifying variants in ", basename(bam)))
                 bamfl <- Rsamtools::BamFile(bam)
                 bpp <- BiocParallel::MulticoreParam(threads)
                 scan <- Rsamtools::scanBam(bam)
                 if (length(scan[[1]]$rname) < 1) next
                 chromosome <- as.character(scan[[1]]$rname[[1]])
                 rm(scan)
                 gr <- GenomicRanges::GRanges(GenomeInfoDb::seqinfo(refgen))
                 gr <- gr[GenomicRanges::seqnames(gr) != chromosome]
                 tally.Param <- VariantTools::TallyVariantsParam(refgen, high_base_quality = 23L,
                                                   mask = gr, indels = TRUE)
                 print("Tallying BAM file...")
                 skip_to_next <- FALSE
                 tryCatch(
                   {
                     tallies <- VariantTools::tallyVariants(bamfl, tally.Param, BPPARAM = bpp)
                   }, error = function(e) {
                     print(paste0("Failed to tally chromosome ", basename(tools::file_path_sans_ext(bam)), "."))
                     skip_to_next <- TRUE
                   })
                 if (skip_to_next == TRUE) {next}
                 gc()
                 calling.filters <- VariantTools::VariantCallingFilters()
                 post.filters <- VariantTools::VariantPostFilters()
                 print("Calling variants...")
                 snp <- VariantTools::callVariants(tallies, calling.filters, post.filters)
                 sampleNames(snp) <- tools::file_path_sans_ext(basename(i))
                 mcols(snp) <- NULL
                 print("Formatting variant information as VCF...")



                 vcfpath <- c()
                 if (length(snp) > 0) {
                   vcf <- VariantAnnotation::asVCF(snp)
                   VariantAnnotation::writeVcf(vcf, paste0(tools::file_path_sans_ext(bam), ".vcf"),
                            index = FALSE)
                   vcf <- VariantAnnotation::readVcf(paste0(tools::file_path_sans_ext(bam), ".vcf"))

                   ##Due to constraints in memory, rsIDs are obtained on a chromosome by chromosome
                   ##basis
                   ##The resulting data frame is sorted as per the vcf and the rsIDs are injected
                   ##into the vcf before writing it as a file again
                   print("Loading dbSNP information...")
                   all_snps <- XtraSNPlocs.Hsapiens.dbSNP144.GRCh37
                   vcfID <- data.frame()
                   mainChromosomes <- paste0("chr", GenomeInfoDb::seqlevels(all_snps))

                   print("Finding rsIDs...")
                   for (x in mainChromosomes) {
                     vcf_chrom <- vcf[grepl(names(vcf),
                                            pattern = paste0(x, ":"))]
                     rd_chr <- rowRanges(vcf_chrom)
                     tar_chr <- as.vector(GenomicRanges::seqnames(rd_chr)@values)
                     if (length(tar_chr) > 0) {
                       tar_chr <- gsub("chr", "", tar_chr)
                       tar_chr <- GenomeInfoDb::seqlevels(all_snps)[GenomeInfoDb::seqlevels(all_snps) == tar_chr]
                       my_snps <- snpsBySeqname(all_snps, c(tar_chr))
                       snp_ID <- data.frame(posIDX = paste0("chr",
                                                            GenomicRanges::seqnames(my_snps),
                                                            ":",
                                                            pos(my_snps)),
                                            rsID = my_snps$RefSNP_id,
                                            stringsAsFactors = FALSE)
                       matV1 <- data.frame(Variant = names(rd_chr), stringsAsFactors = FALSE)
                       matV1$chromosome <- gsub("(.*):(.*)_(.*)/(.*)", "\\1", matV1$Variant)
                       matV1$start <- gsub("(.*):(.*)_(.*)/(.*)", "\\2", matV1$Variant)
                       matV1$end <- gsub("(.*):(.*)_(.*)/(.*)", "\\2", matV1$Variant)
                       matV1$ref_allele <- gsub("(.*):(.*)_(.*)/(.*)", "\\3", matV1$Variant)
                       matV1$alt_allele <- gsub("(.*):(.*)_(.*)/(.*)", "\\4", matV1$Variant)
                       matV1$posIDX <- gsub("(.*)_(.*)", "\\1", matV1$Variant)
                       matV1$start <- as.integer(matV1$start)
                       length <- 1:length(matV1$start)
                       for (i in length) {
                         matV1$end[i] <-
                           matV1$start[i] + (nchar(matV1$ref_allele[i]) - nchar(matV1$alt_allele[i]))
                       }
                       matV1$end[matV1$end < matV1$start] <- matV1$start[matV1$end < matV1$start] - 1
                       matV1$posIDX <- gsub("(.*)_(.*)", "\\1", paste0(matV1$chromosome,
                                                                       ":",
                                                                       matV1$start,
                                                                       "-",
                                                                       matV1$end))
                       matS <- merge(matV1,
                                     snp_ID[!duplicated(snp_ID[, "posIDX"]),],
                                     all.x=TRUE,
                                     by="posIDX")
                       matS <- dplyr::select(matS,-posIDX)
                       vcfID <- merge(vcfID, matS, all.x = TRUE, all.y = TRUE)
                       vcfID <- merge(vcfID, matS, all.x = TRUE, all.y = TRUE)

                       print("Injecting rsIDs into VCF...")
                       vcfID$chr <-  as.integer(gsub("chr", "", vcfID$chromosome))
                       vcfID$chr[vcfID$chromosome == "chrX"] <- 23L
                       vcfID$chr[vcfID$chromosome == "chrY"] <- 24L
                       vcfID$chr[vcfID$chromosome == "chrM"] <- 25L
                       vcfID$chr[vcfID$chromosome == "chrMT"] <- 26L
                       dd <- vcfID[order(vcfID$chr, as.integer(vcfID$start)),]
                       dd$rsID[is.na(dd$rsID)] <- "."
                       rsID <- dd$rsID[dd$chromosome != "chrMT"]
                       if (length(vcfID$chr[vcfID$chromosome == "chrMT"]) > 0) {
                         M <- rep(".", length(names(vcf[GenomicRanges::seqnames(vcf) == "chrM",])))
                         rsID <- append(rsID, M)
                         MT <- dd$rsID[dd$chromosome == "chrMT"]
                         rsID <- append(rsID, MT)
                       }
                       fill <- rep(".", length(names(vcf)) - length(rsID))
                       rsID <- append(rsID, fill)
                       names(vcf) <- rsID
                     }
                   }

                   print("Writing to VCF file...")
                   setwd(tempfolder)
                   print(getwd())
                   print(paste0(tools::file_path_sans_ext(basename(bam)), ".vcf"))
                   a <- VariantAnnotation::writeVcf(vcf, paste0(tools::file_path_sans_ext(basename(bam)), ".vcf"),
                                 index = FALSE)
                   gc()
                   setwd(wd)
                 } else {
                   print(paste0("No variants found in ", (basename(bam))))
                 }
               }

               print("Merging VCF files...")
               vcflist <- tools::list_files_with_exts(tempfolder, "vcf")
               vcfpath <- normalizePath(vcflist)
               match <- VariantAnnotation::readVcf(vcfpath[1])
               print(paste0(gsub(".vcf", "", basename(vcfpath[1])), " OKAY"))
               for (item in vcfpath[-1]) {
                 b <- VariantAnnotation::readVcf(item)
                 match <- BiocGenerics::rbind(match, b)
                 print(paste0(gsub(".vcf", "", basename(item)), " OKAY"))
               }
               rsIDs <- rownames(match)
               rownames(match) <- gsub("chr?(.*)", ".", rsIDs)
               VariantAnnotation::writeVcf(match, paste0(outputdir, "/",
                                      basename(tools::file_path_sans_ext(i)),
                                      "_INDEL", ".vcf"),
                        index = TRUE)
               file.remove(list.files(normalizePath(tempfolder), full.names = TRUE))
               print(paste0("Created ", tools::file_path_sans_ext(basename(i)), "_INDEL", ".vcf.bgz"))
               gc()
               setwd(wd)
               append(output, paste0(outputdir, "/",
                                     tools::file_path_sans_ext(basename(i)),
                                     "_INDEL", ".vcf.bgz"))
             }
             file.remove(normalizePath(tempfolder))
           },
           "2"={
             #library(BSgenome.Hsapiens.UCSC.hg38)
             #library(XtraSNPlocs.Hsapiens.dbSNP144.GRCh38)
             ##Homo sapiens hg38
             if (!requireNamespace("BSgenome.Hsapiens.UCSC.hg38", quietly = TRUE)) {
               stop("The package 'BSgenome.Hsapiens.UCSC.hg38' is required but not installed.
       Install it with: BiocManager::install('BSgenome.Hsapiens.UCSC.hg38')")
             }
             if (!requireNamespace("XtraSNPlocs.Hsapiens.dbSNP144.GRCh38", quietly = TRUE)) {
               stop("The package 'XtraSNPlocs.Hsapiens.dbSNP144.GRCh38' is required but not installed.
       Install it with: BiocManager::install('XtraSNPlocs.Hsapiens.dbSNP144.GRCh38')")
             }
             organism <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38

             ##Selects hg38 as the reference genome
             ##If reference doesn't exist within package directory, create one
             if (dir.exists(paste0(find.package("exvar"), "/hg38"))) {
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"))
             } else {
               print("Reference genome not found. Creating reference. This might take a while...")
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"),
                                    create = TRUE)
             }
             output <- c()
             for (i in bam) {
               print(paste0("Analysing ", basename(i)))
               if (isFALSE(file.exists(paste0(i, ".bai")))) {
                 print("Indexing bam...")
                 setwd(dirname(i))
                 Rsamtools::sortBam(i, tools::file_path_sans_ext(i))
                 Rsamtools::indexBam(i)
               }
               setwd(wd)
               bamfl <- Rsamtools::BamFile(i)
               chromosomes <- GenomeInfoDb::seqlevels(bamfl)
               print("Splitting BAM into chromosomes...")
               tempfolder <- paste0(dirname(i), "/temp")
               dir.create(tempfolder)
               for (x in chromosomes) {
                 print(paste0("Creating ",
                              basename(tools::file_path_sans_ext(i)), "_", x, ".bam ",
                              "in temp folder."))
                 system(paste0("samtools view -b ",
                               i,
                               " ", x, " > ",
                               tempfolder, "/",
                               basename(tools::file_path_sans_ext(i)), "_", x, ".bam"))
                 setwd(tempfolder)
                 Rsamtools::sortBam(paste0(basename(tools::file_path_sans_ext(i)), "_", x, ".bam"),
                         paste0(basename(tools::file_path_sans_ext(i)), "_", x))
                 Rsamtools::indexBam(paste0(basename(tools::file_path_sans_ext(i)), "_", x, ".bam"))
                 setwd(wd)
               }
               chrbam <- tools::list_files_with_exts(tempfolder, "bam")
               chrbam <- sort(chrbam)


               for (bam in chrbam) {
                 print(paste0("Identifying variants in ", basename(bam)))
                 bamfl <- Rsamtools::BamFile(bam)
                 bpp <- BiocParallel::MulticoreParam(threads)
                 scan <- Rsamtools::scanBam(bam)
                 if (length(scan[[1]]$rname) < 1) next
                 chromosome <- as.character(scan[[1]]$rname[[1]])
                 rm(scan)
                 gr <- GenomicRanges::GRanges(GenomeInfoDb::seqinfo(refgen))
                 gr <- gr[GenomicRanges::seqnames(gr) != chromosome]
                 tally.Param <- VariantTools::TallyVariantsParam(refgen, high_base_quality = 23L,
                                                   mask = gr, indels = TRUE)
                 print("Tallying BAM file...")
                 skip_to_next <- FALSE
                 tryCatch(
                   {
                     tallies <- VariantTools::tallyVariants(bamfl, tally.Param, BPPARAM = bpp)
                   }, error = function(e) {
                     print(paste0("Failed to tally chromosome ", basename(tools::file_path_sans_ext(bam)), "."))
                     skip_to_next <- TRUE
                   })
                 if (skip_to_next == TRUE) {next}
                 gc()
                 calling.filters <- VariantTools::VariantCallingFilters()
                 post.filters <- VariantTools::VariantPostFilters()
                 print("Calling variants...")
                 snp <- VariantTools::callVariants(tallies, calling.filters, post.filters)
                 sampleNames(snp) <- tools::file_path_sans_ext(basename(i))
                 mcols(snp) <- NULL
                 print("Formatting variant information as VCF...")



                 vcfpath <- c()
                 if (length(snp) > 0) {
                   vcf <- VariantAnnotation::asVCF(snp)
                   VariantAnnotation::writeVcf(vcf, paste0(tools::file_path_sans_ext(bam), ".vcf"),
                            index = FALSE)
                   vcf <- VariantAnnotation::readVcf(paste0(tools::file_path_sans_ext(bam), ".vcf"))

                   ##Due to constraints in memory, rsIDs are obtained on a chromosome by chromosome
                   ##basis
                   ##The resulting data frame is sorted as per the vcf and the rsIDs are injected
                   ##into the vcf before writing it as a file again
                   print("Loading dbSNP information...")
                   all_snps <- XtraSNPlocs.Hsapiens.dbSNP144.GRCh38
                   vcfID <- data.frame()
                   mainChromosomes <- paste0("chr", GenomeInfoDb::seqlevels(all_snps))

                   print("Finding rsIDs...")
                   for (x in mainChromosomes) {
                     vcf_chrom <- vcf[grepl(names(vcf),
                                            pattern = paste0(x, ":"))]
                     rd_chr <- rowRanges(vcf_chrom)
                     tar_chr <- as.vector(GenomicRanges::seqnames(rd_chr)@values)
                     if (length(tar_chr) > 0) {
                       tar_chr <- gsub("chr", "", tar_chr)
                       tar_chr <- GenomeInfoDb::seqlevels(all_snps)[GenomeInfoDb::seqlevels(all_snps) == tar_chr]
                       my_snps <- snpsBySeqname(all_snps, c(tar_chr))
                       snp_ID <- data.frame(posIDX = paste0("chr",
                                                            GenomicRanges::seqnames(my_snps),
                                                            ":",
                                                            pos(my_snps)),
                                            rsID = my_snps$RefSNP_id,
                                            stringsAsFactors = FALSE)
                       matV1 <- data.frame(Variant = names(rd_chr), stringsAsFactors = FALSE)
                       matV1$chromosome <- gsub("(.*):(.*)_(.*)/(.*)", "\\1", matV1$Variant)
                       matV1$start <- gsub("(.*):(.*)_(.*)/(.*)", "\\2", matV1$Variant)
                       matV1$end <- gsub("(.*):(.*)_(.*)/(.*)", "\\2", matV1$Variant)
                       matV1$ref_allele <- gsub("(.*):(.*)_(.*)/(.*)", "\\3", matV1$Variant)
                       matV1$alt_allele <- gsub("(.*):(.*)_(.*)/(.*)", "\\4", matV1$Variant)
                       matV1$posIDX <- gsub("(.*)_(.*)", "\\1", matV1$Variant)
                       matV1$start <- as.integer(matV1$start)
                       length <- 1:length(matV1$start)
                       for (i in length) {
                         matV1$end[i] <-
                           matV1$start[i] + (nchar(matV1$ref_allele[i]) - nchar(matV1$alt_allele[i]))
                       }
                       matV1$end[matV1$end < matV1$start] <- matV1$start[matV1$end < matV1$start] - 1
                       matV1$posIDX <- gsub("(.*)_(.*)", "\\1", paste0(matV1$chromosome,
                                                                       ":",
                                                                       matV1$start,
                                                                       "-",
                                                                       matV1$end))
                       matS <- merge(matV1,
                                     snp_ID[!duplicated(snp_ID[, "posIDX"]),],
                                     all.x=TRUE,
                                     by="posIDX")
                       matS <- dplyr::select(matS,-posIDX)
                       vcfID <- merge(vcfID, matS, all.x = TRUE, all.y = TRUE)
                       vcfID <- merge(vcfID, matS, all.x = TRUE, all.y = TRUE)

                       print("Injecting rsIDs into VCF...")
                       vcfID$chr <-  as.integer(gsub("chr", "", vcfID$chromosome))
                       vcfID$chr[vcfID$chromosome == "chrX"] <- 23L
                       vcfID$chr[vcfID$chromosome == "chrY"] <- 24L
                       vcfID$chr[vcfID$chromosome == "chrM"] <- 25L
                       vcfID$chr[vcfID$chromosome == "chrMT"] <- 26L
                       dd <- vcfID[order(vcfID$chr, as.integer(vcfID$start)),]
                       dd$rsID[is.na(dd$rsID)] <- "."
                       rsID <- dd$rsID[dd$chromosome != "chrMT"]
                       if (length(vcfID$chr[vcfID$chromosome == "chrMT"]) > 0) {
                         M <- rep(".", length(names(vcf[GenomicRanges::seqnames(vcf) == "chrM",])))
                         rsID <- append(rsID, M)
                         MT <- dd$rsID[dd$chromosome == "chrMT"]
                         rsID <- append(rsID, MT)
                       }
                       fill <- rep(".", length(names(vcf)) - length(rsID))
                       rsID <- append(rsID, fill)
                       names(vcf) <- rsID
                     }
                   }

                   print("Writing to VCF file...")
                   setwd(tempfolder)
                   print(getwd())
                   print(paste0(tools::file_path_sans_ext(basename(bam)), ".vcf"))
                   a <- VariantAnnotation::writeVcf(vcf, paste0(tools::file_path_sans_ext(basename(bam)), ".vcf"),
                                 index = FALSE)
                   gc()
                   setwd(wd)
                 } else {
                   print(paste0("No variants found in ", (basename(bam))))
                 }
               }

               print("Merging VCF files...")
               vcflist <- tools::list_files_with_exts(tempfolder, "vcf")
               vcfpath <- normalizePath(vcflist)
               match <- VariantAnnotation::readVcf(vcfpath[1])
               print(paste0(gsub(".vcf", "", basename(vcfpath[1])), " OKAY"))
               for (item in vcfpath[-1]) {
                 b <- VariantAnnotation::readVcf(item)
                 match <- BiocGenerics::rbind(match, b)
                 print(paste0(gsub(".vcf", "", basename(item)), " OKAY"))
               }
               rsIDs <- rownames(match)
               rownames(match) <- gsub("chr?(.*)", ".", rsIDs)
               VariantAnnotation::writeVcf(match, paste0(outputdir, "/",
                                      basename(tools::file_path_sans_ext(i)),
                                      "_INDEL", ".vcf"),
                        index = TRUE)
               file.remove(list.files(normalizePath(tempfolder), full.names = TRUE))
               print(paste0("Created ", tools::file_path_sans_ext(basename(i)), "_INDEL", ".vcf.bgz"))
               gc()
               setwd(wd)
               append(output, paste0(outputdir, "/",
                                     tools::file_path_sans_ext(basename(i)),
                                     "_INDEL", ".vcf.bgz"))
             }
             file.remove(normalizePath(tempfolder))
           },
           "3"={
             #library(BSgenome.Mmusculus.UCSC.mm10)
             ##Mus musculus mm10
             if (!requireNamespace("BSgenome.Mmusculus.UCSC.mm10", quietly = TRUE)) {
               stop("The package 'BSgenome.Mmusculus.UCSC.mm10' is required but not installed.
       Install it with: BiocManager::install('BSgenome.Mmusculus.UCSC.mm10')")
             }
             organism <- BSgenome.Mmusculus.UCSC.mm10::BSgenome.Mmusculus.UCSC.mm10

             ##Selects mm10 as the reference genome
             ##If reference doesn't exist within package directory, create one
             if (dir.exists(paste0(find.package("exvar"), "/mm10"))) {
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"))
             } else {
               print("Reference genome not found. Creating reference. This might take a while...")
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"),
                                    create = TRUE)
             }
             output <- c()
             for (i in bam) {
               print(paste0("Analysing ", basename(i)))
               if (isFALSE(file.exists(paste0(i, ".bai")))) {
                 print("Indexing bam...")
                 setwd(dirname(i))
                 Rsamtools::sortBam(i, tools::file_path_sans_ext(i))
                 Rsamtools::indexBam(i)
               }
               setwd(wd)
               bamfl <- Rsamtools::BamFile(i)
               chromosomes <- GenomeInfoDb::seqlevels(bamfl)
               print("Splitting BAM into chromosomes...")
               tempfolder <- paste0(dirname(i), "/temp")
               dir.create(tempfolder)
               for (x in chromosomes) {
                 print(paste0("Creating ",
                              basename(tools::file_path_sans_ext(i)), "_", x, ".bam ",
                              "in temp folder."))
                 system(paste0("samtools view -b ",
                               i,
                               " ", x, " > ",
                               tempfolder, "/",
                               basename(tools::file_path_sans_ext(i)), "_", x, ".bam"))
                 setwd(tempfolder)
                 Rsamtools::sortBam(paste0(basename(tools::file_path_sans_ext(i)), "_", x, ".bam"),
                         paste0(basename(tools::file_path_sans_ext(i)), "_", x))
                 Rsamtools::indexBam(paste0(basename(tools::file_path_sans_ext(i)), "_", x, ".bam"))
                 setwd(wd)
               }
               chrbam <- tools::list_files_with_exts(tempfolder, "bam")
               chrbam <- sort(chrbam)


               for (bam in chrbam) {
                 print(paste0("Identifying variants in ", basename(bam)))
                 bamfl <- Rsamtools::BamFile(bam)
                 bpp <- BiocParallel::MulticoreParam(threads)
                 scan <- Rsamtools::scanBam(bam)
                 if (length(scan[[1]]$rname) < 1) next
                 chromosome <- as.character(scan[[1]]$rname[[1]])
                 rm(scan)
                 gr <- GenomicRanges::GRanges(GenomeInfoDb::seqinfo(refgen))
                 gr <- gr[GenomicRanges::seqnames(gr) != chromosome]
                 tally.Param <- VariantTools::TallyVariantsParam(refgen, high_base_quality = 23L,
                                                   mask = gr, indels = TRUE)
                 print("Tallying BAM file...")
                 skip_to_next <- FALSE
                 tryCatch(
                   {
                     tallies <- VariantTools::tallyVariants(bamfl, tally.Param, BPPARAM = bpp)
                   }, error = function(e) {
                     print(paste0("Failed to tally chromosome ", basename(tools::file_path_sans_ext(bam)), "."))
                     skip_to_next <- TRUE
                   })
                 if (skip_to_next == TRUE) {next}
                 gc()
                 calling.filters <- VariantTools::VariantCallingFilters()
                 post.filters <- VariantTools::VariantPostFilters()
                 print("Calling variants...")
                 snp <- VariantTools::callVariants(tallies, calling.filters, post.filters)
                 sampleNames(snp) <- tools::file_path_sans_ext(basename(i))
                 mcols(snp) <- NULL
                 print("Formatting variant information as VCF...")
                 if (length(snp) > 0)  {
                   vcf <- VariantAnnotation::asVCF(snp)
                   print("Writing to VCF file...")
                   setwd(tempfolder)
                   print(getwd())
                   print(paste0(tools::file_path_sans_ext(basename(bam)), ".vcf"))
                   a <- VariantAnnotation::writeVcf(vcf, paste0(tools::file_path_sans_ext(basename(bam)), ".vcf"),
                                 index = FALSE)
                   gc()
                 } else {
                   print(paste0("No variants found in ", (basename(bam))))
                 }

                 setwd(wd)
               }

               print("Merging VCF files...")
               vcflist <- tools::list_files_with_exts(tempfolder, "vcf")
               vcfpath <- normalizePath(vcflist)
               match <- VariantAnnotation::readVcf(vcfpath[1])
               print(paste0(gsub(".vcf", "", basename(vcfpath[1])), " OKAY"))
               for (item in vcfpath[-1]) {
                 b <- VariantAnnotation::readVcf(item)
                 match <- BiocGenerics::rbind(match, b)
                 print(paste0(gsub(".vcf", "", basename(item)), " OKAY"))
               }
               rsIDs <- rownames(match)
               rownames(match) <- gsub("chr?(.*)", ".", rsIDs)
               VariantAnnotation::writeVcf(match, paste0(outputdir, "/",
                                      basename(tools::file_path_sans_ext(i)),
                                      "_INDEL", ".vcf"),
                        index = TRUE)
               file.remove(list.files(normalizePath(tempfolder), full.names = TRUE))
               print(paste0("Created ", tools::file_path_sans_ext(basename(i)), "_INDEL", ".vcf.bgz"))
               gc()
               setwd(wd)
               append(output, paste0(outputdir, "/",
                                     tools::file_path_sans_ext(basename(i)),
                                     "_INDEL", ".vcf.bgz"))
             }
             file.remove(normalizePath(tempfolder))
           },
           "4"={
             #library(BSgenome.Athaliana.TAIR.TAIR9)
             ##Arabidopsis thaliana TAIR9
             if (!requireNamespace("BSgenome.Athaliana.TAIR.TAIR9", quietly = TRUE)) {
               stop("The package 'BSgenome.Athaliana.TAIR.TAIR9' is required but not installed.
       Install it with: BiocManager::install('BSgenome.Athaliana.TAIR.TAIR9')")
             }
             organism <- BSgenome.Athaliana.TAIR.TAIR9::BSgenome.Athaliana.TAIR.TAIR9

             ##Selects hg19 as the reference genome
             ##If reference doesn't exist within package directory, create one
             if (dir.exists(paste0(find.package("exvar"), "/TAIR9"))) {
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"))
             } else {
               print("Reference genome not found. Creating reference. This might take a while...")
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"),
                                    create = TRUE)
             }
             output <- c()
             for (i in bam) {
               print(paste0("Analysing ", basename(i)))
               if (isFALSE(file.exists(paste0(i, ".bai")))) {
                 print("Indexing bam...")
                 setwd(dirname(i))
                 Rsamtools::sortBam(i, tools::file_path_sans_ext(i))
                 Rsamtools::indexBam(i)
               }
               setwd(wd)
               bamfl <- Rsamtools::BamFile(i)
               chromosomes <- GenomeInfoDb::seqlevels(bamfl)
               print("Splitting BAM into chromosomes...")
               tempfolder <- paste0(dirname(i), "/temp")
               dir.create(tempfolder)
               for (x in chromosomes) {
                 print(paste0("Creating ",
                              basename(tools::file_path_sans_ext(i)), "_", x, ".bam ",
                              "in temp folder."))
                 system(paste0("samtools view -b ",
                               i,
                               " ", x, " > ",
                               tempfolder, "/",
                               basename(tools::file_path_sans_ext(i)), "_", x, ".bam"))
                 setwd(tempfolder)
                 Rsamtools::sortBam(paste0(basename(tools::file_path_sans_ext(i)), "_", x, ".bam"),
                         paste0(basename(tools::file_path_sans_ext(i)), "_", x))
                 Rsamtools::indexBam(paste0(basename(tools::file_path_sans_ext(i)), "_", x, ".bam"))
                 setwd(wd)
               }
               chrbam <- tools::list_files_with_exts(tempfolder, "bam")
               chrbam <- sort(chrbam)


               for (bam in chrbam) {
                 print(paste0("Identifying variants in ", basename(bam)))
                 bamfl <- Rsamtools::BamFile(bam)
                 bpp <- BiocParallel::MulticoreParam(threads)
                 scan <- Rsamtools::scanBam(bam)
                 if (length(scan[[1]]$rname) < 1) next
                 chromosome <- as.character(scan[[1]]$rname[[1]])
                 rm(scan)
                 gr <- GenomicRanges::GRanges(GenomeInfoDb::seqinfo(refgen))
                 gr <- gr[GenomicRanges::seqnames(gr) != chromosome]
                 tally.Param <- VariantTools::TallyVariantsParam(refgen, high_base_quality = 23L,
                                                   mask = gr, indels = TRUE)
                 print("Tallying BAM file...")
                 skip_to_next <- FALSE
                 tryCatch(
                   {
                     tallies <- VariantTools::tallyVariants(bamfl, tally.Param, BPPARAM = bpp)
                   }, error = function(e) {
                     print(paste0("Failed to tally chromosome ", basename(tools::file_path_sans_ext(bam)), "."))
                     skip_to_next <- TRUE
                   })
                 if (skip_to_next == TRUE) {next}
                 gc()
                 calling.filters <- VariantTools::VariantCallingFilters()
                 post.filters <- VariantTools::VariantPostFilters()
                 print("Calling variants...")
                 snp <- VariantTools::callVariants(tallies, calling.filters, post.filters)
                 sampleNames(snp) <- tools::file_path_sans_ext(basename(i))
                 mcols(snp) <- NULL
                 print("Formatting variant information as VCF...")
                 if (length(snp) > 0)  {
                   vcf <- VariantAnnotation::asVCF(snp)
                   print("Writing to VCF file...")
                   setwd(tempfolder)
                   print(getwd())
                   print(paste0(tools::file_path_sans_ext(basename(bam)), ".vcf"))
                   a <- VariantAnnotation::writeVcf(vcf, paste0(tools::file_path_sans_ext(basename(bam)), ".vcf"),
                                 index = FALSE)
                   gc()
                 } else {
                   print(paste0("No variants found in ", (basename(bam))))
                 }

                 setwd(wd)
               }

               print("Merging VCF files...")
               vcflist <- tools::list_files_with_exts(tempfolder, "vcf")
               vcfpath <- normalizePath(vcflist)
               match <- VariantAnnotation::readVcf(vcfpath[1])
               print(paste0(gsub(".vcf", "", basename(vcfpath[1])), " OKAY"))
               for (item in vcfpath[-1]) {
                 b <- VariantAnnotation::readVcf(item)
                 match <- BiocGenerics::rbind(match, b)
                 print(paste0(gsub(".vcf", "", basename(item)), " OKAY"))
               }
               rsIDs <- rownames(match)
               rownames(match) <- gsub("chr?(.*)", ".", rsIDs)
               VariantAnnotation::writeVcf(match, paste0(outputdir, "/",
                                      basename(tools::file_path_sans_ext(i)),
                                      "_INDEL", ".vcf"),
                        index = TRUE)
               file.remove(list.files(normalizePath(tempfolder), full.names = TRUE))
               print(paste0("Created ", tools::file_path_sans_ext(basename(i)), "_INDEL", ".vcf.bgz"))
               gc()
               setwd(wd)
               append(output, paste0(outputdir, "/",
                                     tools::file_path_sans_ext(basename(i)),
                                     "_INDEL", ".vcf.bgz"))
             }
             file.remove(normalizePath(tempfolder))
           },
           "5"={
             #library(BSgenome.Dmelanogaster.UCSC.dm6)
             ##Drosophilia melanogaster dm6
             if (!requireNamespace("BSgenome.Dmelanogaster.UCSC.dm6", quietly = TRUE)) {
               stop("The package 'BSgenome.Dmelanogaster.UCSC.dm6' is required but not installed.
       Install it with: BiocManager::install('BSgenome.Dmelanogaster.UCSC.dm6')")
             }
             organism <- BSgenome.Dmelanogaster.UCSC.dm6::BSgenome.Dmelanogaster.UCSC.dm6

             ##Selects dm6 as the reference genome
             ##If reference doesn't exist within package directory, create one
             if (dir.exists(paste0(find.package("exvar"), "/dm6"))) {
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"))
             } else {
               print("Reference genome not found. Creating reference. This might take a while...")
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"),
                                    create = TRUE)
             }
             output <- c()
             for (i in bam) {
               print(paste0("Analysing ", basename(i)))
               if (isFALSE(file.exists(paste0(i, ".bai")))) {
                 print("Indexing bam...")
                 setwd(dirname(i))
                 Rsamtools::sortBam(i, tools::file_path_sans_ext(i))
                 Rsamtools::indexBam(i)
               }
               setwd(wd)
               bamfl <- Rsamtools::BamFile(i)
               chromosomes <- GenomeInfoDb::seqlevels(bamfl)
               print("Splitting BAM into chromosomes...")
               tempfolder <- paste0(dirname(i), "/temp")
               dir.create(tempfolder)
               for (x in chromosomes) {
                 print(paste0("Creating ",
                              basename(tools::file_path_sans_ext(i)), "_", x, ".bam ",
                              "in temp folder."))
                 system(paste0("samtools view -b ",
                               i,
                               " ", x, " > ",
                               tempfolder, "/",
                               basename(tools::file_path_sans_ext(i)), "_", x, ".bam"))
                 setwd(tempfolder)
                 Rsamtools::sortBam(paste0(basename(tools::file_path_sans_ext(i)), "_", x, ".bam"),
                         paste0(basename(tools::file_path_sans_ext(i)), "_", x))
                 Rsamtools::indexBam(paste0(basename(tools::file_path_sans_ext(i)), "_", x, ".bam"))
                 setwd(wd)
               }
               chrbam <- tools::list_files_with_exts(tempfolder, "bam")
               chrbam <- sort(chrbam)


               for (bam in chrbam) {
                 print(paste0("Identifying variants in ", basename(bam)))
                 bamfl <- Rsamtools::BamFile(bam)
                 bpp <- BiocParallel::MulticoreParam(threads)
                 scan <- Rsamtools::scanBam(bam)
                 if (length(scan[[1]]$rname) < 1) next
                 chromosome <- as.character(scan[[1]]$rname[[1]])
                 rm(scan)
                 gr <- GenomicRanges::GRanges(GenomeInfoDb::seqinfo(refgen))
                 gr <- gr[GenomicRanges::seqnames(gr) != chromosome]
                 tally.Param <- VariantTools::TallyVariantsParam(refgen, high_base_quality = 23L,
                                                   mask = gr, indels = TRUE)
                 print("Tallying BAM file...")
                 skip_to_next <- FALSE
                 tryCatch(
                   {
                     tallies <- VariantTools::tallyVariants(bamfl, tally.Param, BPPARAM = bpp)
                   }, error = function(e) {
                     print(paste0("Failed to tally chromosome ", basename(tools::file_path_sans_ext(bam)), "."))
                     skip_to_next <- TRUE
                   })
                 if (skip_to_next == TRUE) {next}
                 gc()
                 calling.filters <- VariantTools::VariantCallingFilters()
                 post.filters <- VariantTools::VariantPostFilters()
                 print("Calling variants...")
                 snp <- VariantTools::callVariants(tallies, calling.filters, post.filters)
                 sampleNames(snp) <- tools::file_path_sans_ext(basename(i))
                 mcols(snp) <- NULL
                 print("Formatting variant information as VCF...")
                 if (length(snp) > 0)  {
                   vcf <- VariantAnnotation::asVCF(snp)
                   print("Writing to VCF file...")
                   setwd(tempfolder)
                   print(getwd())
                   print(paste0(tools::file_path_sans_ext(basename(bam)), ".vcf"))
                   a <- VariantAnnotation::writeVcf(vcf, paste0(tools::file_path_sans_ext(basename(bam)), ".vcf"),
                                 index = FALSE)
                   gc()
                 } else {
                   print(paste0("No variants found in ", (basename(bam))))
                 }

                 setwd(wd)
               }

               print("Merging VCF files...")
               vcflist <- tools::list_files_with_exts(tempfolder, "vcf")
               vcfpath <- normalizePath(vcflist)
               match <- VariantAnnotation::readVcf(vcfpath[1])
               print(paste0(gsub(".vcf", "", basename(vcfpath[1])), " OKAY"))
               for (item in vcfpath[-1]) {
                 b <- VariantAnnotation::readVcf(item)
                 match <- BiocGenerics::rbind(match, b)
                 print(paste0(gsub(".vcf", "", basename(item)), " OKAY"))
               }
               rsIDs <- rownames(match)
               rownames(match) <- gsub("chr?(.*)", ".", rsIDs)
               VariantAnnotation::writeVcf(match, paste0(outputdir, "/",
                                      basename(tools::file_path_sans_ext(i)),
                                      "_INDEL", ".vcf"),
                        index = TRUE)
               file.remove(list.files(normalizePath(tempfolder), full.names = TRUE))
               print(paste0("Created ", tools::file_path_sans_ext(basename(i)), "_INDEL", ".vcf.bgz"))
               gc()
               setwd(wd)
               append(output, paste0(outputdir, "/",
                                     tools::file_path_sans_ext(basename(i)),
                                     "_INDEL", ".vcf.bgz"))
             }
             file.remove(normalizePath(tempfolder))
           },
           "6"={
             #library(BSgenome.Drerio.UCSC.danRer11)
             ##Danio rerio danRer11
             if (!requireNamespace("BSgenome.Drerio.UCSC.danRer11", quietly = TRUE)) {
               stop("The package 'BSgenome.Drerio.UCSC.danRer11' is required but not installed.
       Install it with: BiocManager::install('BSgenome.Drerio.UCSC.danRer11')")
             }
             organism <- BSgenome.Drerio.UCSC.danRer11::BSgenome.Drerio.UCSC.danRer11

             ##Selects danRer11 as the reference genome
             ##If reference doesn't exist within package directory, create one
             if (dir.exists(paste0(find.package("exvar"), "/danRer11"))) {
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"))
             } else {
               print("Reference genome not found. Creating reference. This might take a while...")
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"),
                                    create = TRUE)
             }
             output <- c()
             for (i in bam) {
               print(paste0("Analysing ", basename(i)))
               if (isFALSE(file.exists(paste0(i, ".bai")))) {
                 print("Indexing bam...")
                 setwd(dirname(i))
                 Rsamtools::sortBam(i, tools::file_path_sans_ext(i))
                 Rsamtools::indexBam(i)
               }
               setwd(wd)
               bamfl <- Rsamtools::BamFile(i)
               chromosomes <- GenomeInfoDb::seqlevels(bamfl)
               print("Splitting BAM into chromosomes...")
               tempfolder <- paste0(dirname(i), "/temp")
               dir.create(tempfolder)
               for (x in chromosomes) {
                 print(paste0("Creating ",
                              basename(tools::file_path_sans_ext(i)), "_", x, ".bam ",
                              "in temp folder."))
                 system(paste0("samtools view -b ",
                               i,
                               " ", x, " > ",
                               tempfolder, "/",
                               basename(tools::file_path_sans_ext(i)), "_", x, ".bam"))
                 setwd(tempfolder)
                 Rsamtools::sortBam(paste0(basename(tools::file_path_sans_ext(i)), "_", x, ".bam"),
                         paste0(basename(tools::file_path_sans_ext(i)), "_", x))
                 Rsamtools::indexBam(paste0(basename(tools::file_path_sans_ext(i)), "_", x, ".bam"))
                 setwd(wd)
               }
               chrbam <- tools::list_files_with_exts(tempfolder, "bam")
               chrbam <- sort(chrbam)


               for (bam in chrbam) {
                 print(paste0("Identifying variants in ", basename(bam)))
                 bamfl <- Rsamtools::BamFile(bam)
                 bpp <- BiocParallel::MulticoreParam(threads)
                 scan <- Rsamtools::scanBam(bam)
                 if (length(scan[[1]]$rname) < 1) next
                 chromosome <- as.character(scan[[1]]$rname[[1]])
                 rm(scan)
                 gr <- GenomicRanges::GRanges(GenomeInfoDb::seqinfo(refgen))
                 gr <- gr[GenomicRanges::seqnames(gr) != chromosome]
                 tally.Param <- VariantTools::TallyVariantsParam(refgen, high_base_quality = 23L,
                                                   mask = gr, indels = TRUE)
                 print("Tallying BAM file...")
                 skip_to_next <- FALSE
                 tryCatch(
                   {
                     tallies <- VariantTools::tallyVariants(bamfl, tally.Param, BPPARAM = bpp)
                   }, error = function(e) {
                     print(paste0("Failed to tally chromosome ", basename(tools::file_path_sans_ext(bam)), "."))
                     skip_to_next <- TRUE
                   })
                 if (skip_to_next == TRUE) {next}
                 gc()
                 calling.filters <- VariantTools::VariantCallingFilters()
                 post.filters <- VariantTools::VariantPostFilters()
                 print("Calling variants...")
                 snp <- VariantTools::callVariants(tallies, calling.filters, post.filters)
                 sampleNames(snp) <- tools::file_path_sans_ext(basename(i))
                 mcols(snp) <- NULL
                 print("Formatting variant information as VCF...")
                 if (length(snp) > 0)  {
                   vcf <- VariantAnnotation::asVCF(snp)
                   print("Writing to VCF file...")
                   setwd(tempfolder)
                   print(getwd())
                   print(paste0(tools::file_path_sans_ext(basename(bam)), ".vcf"))
                   a <- VariantAnnotation::writeVcf(vcf, paste0(tools::file_path_sans_ext(basename(bam)), ".vcf"),
                                 index = FALSE)
                   gc()
                 } else {
                   print(paste0("No variants found in ", (basename(bam))))
                 }

                 setwd(wd)
               }

               print("Merging VCF files...")
               vcflist <- tools::list_files_with_exts(tempfolder, "vcf")
               vcfpath <- normalizePath(vcflist)
               match <- VariantAnnotation::readVcf(vcfpath[1])
               print(paste0(gsub(".vcf", "", basename(vcfpath[1])), " OKAY"))
               for (item in vcfpath[-1]) {
                 b <- VariantAnnotation::readVcf(item)
                 match <- BiocGenerics::rbind(match, b)
                 print(paste0(gsub(".vcf", "", basename(item)), " OKAY"))
               }
               rsIDs <- rownames(match)
               rownames(match) <- gsub("chr?(.*)", ".", rsIDs)
               VariantAnnotation::writeVcf(match, paste0(outputdir, "/",
                                      basename(tools::file_path_sans_ext(i)),
                                      "_INDEL", ".vcf"),
                        index = TRUE)
               file.remove(list.files(normalizePath(tempfolder), full.names = TRUE))
               print(paste0("Created ", tools::file_path_sans_ext(basename(i)), "_INDEL", ".vcf.bgz"))
               gc()
               setwd(wd)
               append(output, paste0(outputdir, "/",
                                     tools::file_path_sans_ext(basename(i)),
                                     "_INDEL", ".vcf.bgz"))
             }
             file.remove(normalizePath(tempfolder))
           },
           "7"={
             #library(BSgenome.Rnorvegicus.UCSC.rn5)
             ##Rattus norvegicus rn5
             if (!requireNamespace("BSgenome.Rnorvegicus.UCSC.rn5", quietly = TRUE)) {
               stop("The package 'BSgenome.Rnorvegicus.UCSC.rn5' is required but not installed.
       Install it with: BiocManager::install('BSgenome.Rnorvegicus.UCSC.rn5')")
             }
             organism <- BSgenome.Rnorvegicus.UCSC.rn5::BSgenome.Rnorvegicus.UCSC.rn5

             ##Selects danRer11 as the reference genome
             ##If reference doesn't exist within package directory, create one
             if (dir.exists(paste0(find.package("exvar"), "/rn5"))) {
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"))
             } else {
               print("Reference genome not found. Creating reference. This might take a while...")
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"),
                                    create = TRUE)
             }
             output <- c()
             for (i in bam) {
               print(paste0("Analysing ", basename(i)))
               if (isFALSE(file.exists(paste0(i, ".bai")))) {
                 print("Indexing bam...")
                 setwd(dirname(i))
                 Rsamtools::sortBam(i, tools::file_path_sans_ext(i))
                 Rsamtools::indexBam(i)
               }
               setwd(wd)
               bamfl <- Rsamtools::BamFile(i)
               chromosomes <- GenomeInfoDb::seqlevels(bamfl)
               print("Splitting BAM into chromosomes...")
               tempfolder <- paste0(dirname(i), "/temp")
               dir.create(tempfolder)
               for (x in chromosomes) {
                 print(paste0("Creating ",
                              basename(tools::file_path_sans_ext(i)), "_", x, ".bam ",
                              "in temp folder."))
                 system(paste0("samtools view -b ",
                               i,
                               " ", x, " > ",
                               tempfolder, "/",
                               basename(tools::file_path_sans_ext(i)), "_", x, ".bam"))
                 setwd(tempfolder)
                 Rsamtools::sortBam(paste0(basename(tools::file_path_sans_ext(i)), "_", x, ".bam"),
                         paste0(basename(tools::file_path_sans_ext(i)), "_", x))
                 Rsamtools::indexBam(paste0(basename(tools::file_path_sans_ext(i)), "_", x, ".bam"))
                 setwd(wd)
               }
               chrbam <- tools::list_files_with_exts(tempfolder, "bam")
               chrbam <- sort(chrbam)


               for (bam in chrbam) {
                 print(paste0("Identifying variants in ", basename(bam)))
                 bamfl <- Rsamtools::BamFile(bam)
                 bpp <- BiocParallel::MulticoreParam(threads)
                 scan <- Rsamtools::scanBam(bam)
                 if (length(scan[[1]]$rname) < 1) next
                 chromosome <- as.character(scan[[1]]$rname[[1]])
                 rm(scan)
                 gr <- GenomicRanges::GRanges(GenomeInfoDb::seqinfo(refgen))
                 gr <- gr[GenomicRanges::seqnames(gr) != chromosome]
                 tally.Param <- VariantTools::TallyVariantsParam(refgen, high_base_quality = 23L,
                                                   mask = gr, indels = TRUE)
                 print("Tallying BAM file...")
                 skip_to_next <- FALSE
                 tryCatch(
                   {
                     tallies <- VariantTools::tallyVariants(bamfl, tally.Param, BPPARAM = bpp)
                   }, error = function(e) {
                     print(paste0("Failed to tally chromosome ", basename(tools::file_path_sans_ext(bam)), "."))
                     skip_to_next <- TRUE
                   })
                 if (skip_to_next == TRUE) {next}
                 gc()
                 calling.filters <- VariantTools::VariantCallingFilters()
                 post.filters <- VariantTools::VariantPostFilters()
                 print("Calling variants...")
                 snp <- VariantTools::callVariants(tallies, calling.filters, post.filters)
                 sampleNames(snp) <- tools::file_path_sans_ext(basename(i))
                 mcols(snp) <- NULL
                 print("Formatting variant information as VCF...")
                 if (length(snp) > 0)  {
                   vcf <- VariantAnnotation::asVCF(snp)
                   print("Writing to VCF file...")
                   setwd(tempfolder)
                   print(getwd())
                   print(paste0(tools::file_path_sans_ext(basename(bam)), ".vcf"))
                   a <- VariantAnnotation::writeVcf(vcf, paste0(tools::file_path_sans_ext(basename(bam)), ".vcf"),
                                 index = FALSE)
                   gc()
                 } else {
                   print(paste0("No variants found in ", (basename(bam))))
                 }

                 setwd(wd)
               }

               print("Merging VCF files...")
               vcflist <- tools::list_files_with_exts(tempfolder, "vcf")
               vcfpath <- normalizePath(vcflist)
               match <- VariantAnnotation::readVcf(vcfpath[1])
               print(paste0(gsub(".vcf", "", basename(vcfpath[1])), " OKAY"))
               for (item in vcfpath[-1]) {
                 b <- VariantAnnotation::readVcf(item)
                 match <- BiocGenerics::rbind(match, b)
                 print(paste0(gsub(".vcf", "", basename(item)), " OKAY"))
               }
               rsIDs <- rownames(match)
               rownames(match) <- gsub("chr?(.*)", ".", rsIDs)
               VariantAnnotation::writeVcf(match, paste0(outputdir, "/",
                                      basename(tools::file_path_sans_ext(i)),
                                      "_INDEL", ".vcf"),
                        index = TRUE)
               file.remove(list.files(normalizePath(tempfolder), full.names = TRUE))
               print(paste0("Created ", tools::file_path_sans_ext(basename(i)), "_INDEL", ".vcf.bgz"))
               gc()
               setwd(wd)
               append(output, paste0(outputdir, "/",
                                     tools::file_path_sans_ext(basename(i)),
                                     "_INDEL", ".vcf.bgz"))
             }
             file.remove(normalizePath(tempfolder))
           },
           "8"={
             #library(BSgenome.Scerevisiae.UCSC.sacCer3)
             ##Saccharomyces cerevisiae sacCer3
             if (!requireNamespace("BSgenome.Scerevisiae.UCSC.sacCer3", quietly = TRUE)) {
               stop("The package 'BSgenome.Scerevisiae.UCSC.sacCer3' is required but not installed.
       Install it with: BiocManager::install('BSgenome.Scerevisiae.UCSC.sacCer3')")
             }
             organism <- BSgenome.Scerevisiae.UCSC.sacCer3::BSgenome.Scerevisiae.UCSC.sacCer3

             ##Selects sacCer3 as the reference genome
             ##If reference doesn't exist within package directory, create one
             if (dir.exists(paste0(find.package("exvar"), "/sacCer3"))) {
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"))
             } else {
               print("Reference genome not found. Creating reference. This might take a while...")
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"),
                                    create = TRUE)
             }
             output <- c()
             for (i in bam) {
               print(paste0("Analysing ", basename(i)))
               if (isFALSE(file.exists(paste0(i, ".bai")))) {
                 print("Indexing bam...")
                 setwd(dirname(i))
                 Rsamtools::sortBam(i, tools::file_path_sans_ext(i))
                 Rsamtools::indexBam(i)
               }
               setwd(wd)
               bamfl <- Rsamtools::BamFile(i)
               chromosomes <- GenomeInfoDb::seqlevels(bamfl)
               print("Splitting BAM into chromosomes...")
               tempfolder <- paste0(dirname(i), "/temp")
               dir.create(tempfolder)
               for (x in chromosomes) {
                 print(paste0("Creating ",
                              basename(tools::file_path_sans_ext(i)), "_", x, ".bam ",
                              "in temp folder."))
                 system(paste0("samtools view -b ",
                               i,
                               " ", x, " > ",
                               tempfolder, "/",
                               basename(tools::file_path_sans_ext(i)), "_", x, ".bam"))
                 setwd(tempfolder)
                 Rsamtools::sortBam(paste0(basename(tools::file_path_sans_ext(i)), "_", x, ".bam"),
                         paste0(basename(tools::file_path_sans_ext(i)), "_", x))
                 Rsamtools::indexBam(paste0(basename(tools::file_path_sans_ext(i)), "_", x, ".bam"))
                 setwd(wd)
               }
               chrbam <- tools::list_files_with_exts(tempfolder, "bam")
               chrbam <- sort(chrbam)


               for (bam in chrbam) {
                 print(paste0("Identifying variants in ", basename(bam)))
                 bamfl <- Rsamtools::BamFile(bam)
                 bpp <- BiocParallel::MulticoreParam(threads)
                 scan <- Rsamtools::scanBam(bam)
                 if (length(scan[[1]]$rname) < 1) next
                 chromosome <- as.character(scan[[1]]$rname[[1]])
                 rm(scan)
                 gr <- GenomicRanges::GRanges(GenomeInfoDb::seqinfo(refgen))
                 gr <- gr[GenomicRanges::seqnames(gr) != chromosome]
                 tally.Param <- VariantTools::TallyVariantsParam(refgen, high_base_quality = 23L,
                                                   mask = gr, indels = TRUE)
                 print("Tallying BAM file...")
                 skip_to_next <- FALSE
                 tryCatch(
                   {
                     tallies <- VariantTools::tallyVariants(bamfl, tally.Param, BPPARAM = bpp)
                   }, error = function(e) {
                     print(paste0("Failed to tally chromosome ", basename(tools::file_path_sans_ext(bam)), "."))
                     skip_to_next <- TRUE
                   })
                 if (skip_to_next == TRUE) {next}
                 gc()
                 calling.filters <- VariantTools::VariantCallingFilters()
                 post.filters <- VariantTools::VariantPostFilters()
                 print("Calling variants...")
                 snp <- VariantTools::callVariants(tallies, calling.filters, post.filters)
                 sampleNames(snp) <- tools::file_path_sans_ext(basename(i))
                 mcols(snp) <- NULL
                 print("Formatting variant information as VCF...")
                 if (length(snp) > 0)  {
                   vcf <- VariantAnnotation::asVCF(snp)
                   print("Writing to VCF file...")
                   setwd(tempfolder)
                   print(getwd())
                   print(paste0(tools::file_path_sans_ext(basename(bam)), ".vcf"))
                   a <- VariantAnnotation::writeVcf(vcf, paste0(tools::file_path_sans_ext(basename(bam)), ".vcf"),
                                 index = FALSE)
                   gc()
                 } else {
                   print(paste0("No variants found in ", (basename(bam))))
                 }

                 setwd(wd)
               }

               print("Merging VCF files...")
               vcflist <- tools::list_files_with_exts(tempfolder, "vcf")
               vcfpath <- normalizePath(vcflist)
               match <- VariantAnnotation::readVcf(vcfpath[1])
               print(paste0(gsub(".vcf", "", basename(vcfpath[1])), " OKAY"))
               for (item in vcfpath[-1]) {
                 b <- VariantAnnotation::readVcf(item)
                 match <- BiocGenerics::rbind(match, b)
                 print(paste0(gsub(".vcf", "", basename(item)), " OKAY"))
               }
               rsIDs <- rownames(match)
               rownames(match) <- gsub("chr?(.*)", ".", rsIDs)
               VariantAnnotation::writeVcf(match, paste0(outputdir, "/",
                                      basename(tools::file_path_sans_ext(i)),
                                      "_INDEL", ".vcf"),
                        index = TRUE)
               file.remove(list.files(normalizePath(tempfolder), full.names = TRUE))
               print(paste0("Created ", tools::file_path_sans_ext(basename(i)), "_INDEL", ".vcf.bgz"))
               gc()
               setwd(wd)
               append(output, paste0(outputdir, "/",
                                     tools::file_path_sans_ext(basename(i)),
                                     "_INDEL", ".vcf.bgz"))
             }
             file.remove(normalizePath(tempfolder))
           },
           "9"={
             library(BSgenome.Celegans.UCSC.ce11)
             ##Caenorhabditis elagans
             if (!requireNamespace("BSgenome.Celegans.UCSC.ce11", quietly = TRUE)) {
               stop("The package 'BSgenome.Celegans.UCSC.ce11' is required but not installed.
       Install it with: BiocManager::install('BSgenome.Celegans.UCSC.ce11')")
             }
             organism <- BSgenome.Celegans.UCSC.ce11::BSgenome.Celegans.UCSC.ce11

             ##Selects ce11 as the reference genome
             ##If reference doesn't exist within package directory, create one
             if (dir.exists(paste0(find.package("exvar"), "/ce11"))) {
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"))
             } else {
               print("Reference genome not found. Creating reference. This might take a while...")
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"),
                                    create = TRUE)
             }
             output <- c()
             for (i in bam) {
               print(paste0("Analysing ", basename(i)))
               if (isFALSE(file.exists(paste0(i, ".bai")))) {
                 print("Indexing bam...")
                 setwd(dirname(i))
                 Rsamtools::sortBam(i, tools::file_path_sans_ext(i))
                 Rsamtools::indexBam(i)
               }
               setwd(wd)
               bamfl <- Rsamtools::BamFile(i)
               chromosomes <- GenomeInfoDb::seqlevels(bamfl)
               print("Splitting BAM into chromosomes...")
               tempfolder <- paste0(dirname(i), "/temp")
               dir.create(tempfolder)
               for (x in chromosomes) {
                 print(paste0("Creating ",
                              basename(tools::file_path_sans_ext(i)), "_", x, ".bam ",
                              "in temp folder."))
                 system(paste0("samtools view -b ",
                               i,
                               " ", x, " > ",
                               tempfolder, "/",
                               basename(tools::file_path_sans_ext(i)), "_", x, ".bam"))
                 setwd(tempfolder)
                 Rsamtools::sortBam(paste0(basename(tools::file_path_sans_ext(i)), "_", x, ".bam"),
                         paste0(basename(tools::file_path_sans_ext(i)), "_", x))
                 Rsamtools::indexBam(paste0(basename(tools::file_path_sans_ext(i)), "_", x, ".bam"))
                 setwd(wd)
               }
               chrbam <- tools::list_files_with_exts(tempfolder, "bam")
               chrbam <- sort(chrbam)


               for (bam in chrbam) {
                 print(paste0("Identifying variants in ", basename(bam)))
                 bamfl <- Rsamtools::BamFile(bam)
                 bpp <- BiocParallel::MulticoreParam(threads)
                 scan <- Rsamtools::scanBam(bam)
                 if (length(scan[[1]]$rname) < 1) next
                 chromosome <- as.character(scan[[1]]$rname[[1]])
                 rm(scan)
                 gr <- GenomicRanges::GRanges(GenomeInfoDb::seqinfo(refgen))
                 gr <- gr[GenomicRanges::seqnames(gr) != chromosome]
                 tally.Param <- VariantTools::TallyVariantsParam(refgen, high_base_quality = 23L,
                                                   mask = gr, indels = TRUE)
                 print("Tallying BAM file...")
                 skip_to_next <- FALSE
                 tryCatch(
                   {
                     tallies <- VariantTools::tallyVariants(bamfl, tally.Param, BPPARAM = bpp)
                   }, error = function(e) {
                     print(paste0("Failed to tally chromosome ", basename(tools::file_path_sans_ext(bam)), "."))
                     skip_to_next <- TRUE
                   })
                 if (skip_to_next == TRUE) {next}
                 gc()
                 calling.filters <- VariantTools::VariantCallingFilters()
                 post.filters <- VariantTools::VariantPostFilters()
                 print("Calling variants...")
                 snp <- VariantTools::callVariants(tallies, calling.filters, post.filters)
                 sampleNames(snp) <- tools::file_path_sans_ext(basename(i))
                 mcols(snp) <- NULL
                 print("Formatting variant information as VCF...")
                 if (length(snp) > 0)  {
                   vcf <- VariantAnnotation::asVCF(snp)
                   print("Writing to VCF file...")
                   setwd(tempfolder)
                   print(getwd())
                   print(paste0(tools::file_path_sans_ext(basename(bam)), ".vcf"))
                   a <- VariantAnnotation::writeVcf(vcf, paste0(tools::file_path_sans_ext(basename(bam)), ".vcf"),
                                 index = FALSE)
                   gc()
                 } else {
                   print(paste0("No variants found in ", (basename(bam))))
                 }

                 setwd(wd)
               }

               print("Merging VCF files...")
               vcflist <- tools::list_files_with_exts(tempfolder, "vcf")
               vcfpath <- normalizePath(vcflist)
               match <- VariantAnnotation::readVcf(vcfpath[1])
               print(paste0(gsub(".vcf", "", basename(vcfpath[1])), " OKAY"))
               for (item in vcfpath[-1]) {
                 b <- VariantAnnotation::readVcf(item)
                 match <- BiocGenerics::rbind(match, b)
                 print(paste0(gsub(".vcf", "", basename(item)), " OKAY"))
               }
               rsIDs <- rownames(match)
               rownames(match) <- gsub("chr?(.*)", ".", rsIDs)
               VariantAnnotation::writeVcf(match, paste0(outputdir, "/",
                                      basename(tools::file_path_sans_ext(i)),
                                      "_INDEL", ".vcf"),
                        index = TRUE)
               file.remove(list.files(normalizePath(tempfolder), full.names = TRUE))
               print(paste0("Created ", tools::file_path_sans_ext(basename(i)), "_INDEL", ".vcf.bgz"))
               gc()
               setwd(wd)
               append(output, paste0(outputdir, "/",
                                     tools::file_path_sans_ext(basename(i)),
                                     "_INDEL", ".vcf.bgz"))
             }
             file.remove(normalizePath(tempfolder))
           }
    )
  } else if(mode == "n") {
    ##Sets the reference genome that corresponds to the species chosen by the user
    switch(species,
           "1"={
             ##Homo sapiens hg19
             #library(BSgenome.Hsapiens.UCSC.hg19)
             #library(SNPlocs.Hsapiens.dbSNP144.GRCh37) # typo?
             if (!requireNamespace("BSgenome.Hsapiens.UCSC.hg19", quietly = TRUE)) {
               stop("The package 'BSgenome.Hsapiens.UCSC.hg19' is required but not installed.
       Install it with: BiocManager::install('BSgenome.Hsapiens.UCSC.hg19')")
             }
             if (!requireNamespace("XtraSNPlocs.Hsapiens.dbSNP144.GRCh37", quietly = TRUE)) {
               stop("The package 'XtraSNPlocs.Hsapiens.dbSNP144.GRCh37' is required but not installed.
       Install it with: BiocManager::install('XtraSNPlocs.Hsapiens.dbSNP144.GRCh37')")
             }
             organism <- BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19

             ##Selects hg19 as the reference genome
             ##If reference doesn't exist within package directory, create one
             if (dir.exists(paste0(find.package("exvar"), "/hg19"))) {
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"))
             } else {
               print("Reference genome not found. Creating reference. This might take a while...")
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"),
                                    create = TRUE)
             }

             output <- c()
             for(i in bam) {
               print(paste0("Analysing ", basename(i)))
               if (isFALSE(file.exists(paste0(i, ".bai")))) {
                 print("Indexing bam...")
                 setwd(dirname(i))
                 Rsamtools::sortBam(i, tools::file_path_sans_ext(i))
                 Rsamtools::indexBam(i)
               }
               tempfolder <- paste0(dirname(i), "/temp")
               dir.create(tempfolder)
               vcflist <- c()
               bpp <- BiocParallel::MulticoreParam(threads)
               chromosomes <- GenomeInfoDb::seqlevels(refgen)
               bamfl <- Rsamtools::BamFile(i)
               tally.Param <- VariantTools::TallyVariantsParam(refgen, high_base_quality = 23L, indels = TRUE)
               print("Tallying BAM file...")
               tallies <- VariantTools::tallyVariants(bamfl, tally.Param, BPPARAM = bpp)
               gc()
               calling.filters <- VariantTools::VariantCallingFilters()
               post.filters <- VariantTools::VariantPostFilters()
               print("Calling variants...")
               snp <- VariantTools::callVariants(tallies, calling.filters, post.filters)
               sampleNames(snp) <- tools::file_path_sans_ext(dirname(i))
               mcols(snp) <- NULL
               print("Formatting variant information as VCF...")
               vcf <- VariantAnnotation::asVCF(snp)
               VariantAnnotation::writeVcf(vcf, paste0(tempfolder, "/",
                                    basename(tools::file_path_sans_ext(i)),
                                    "_INDEL", ".vcf"))


               vcflist <- c()
               vcf <- VariantAnnotation::readVcf(paste0(tempfolder, "/",
                                     basename(tools::file_path_sans_ext(i)),
                                     "_INDEL", ".vcf"))

               ##Due to constraints in memory, rsIDs are obtained on a chromosome by chromosome
               ##basis
               ##The resulting data frame is sorted as per the vcf and the rsIDs are injected
               ##into the vcf before writing it as a file again
               print("Loading dbSNP information...")
               all_snps <- XtraSNPlocs.Hsapiens.dbSNP144.GRCh37
               vcfID <- data.frame()
               mainChromosomes <- paste0("chr", GenomeInfoDb::seqlevels(all_snps))

               print("Finding rsIDs...")
               for (x in mainChromosomes) {
                 vcf_chrom <- vcf[grepl(names(vcf),
                                        pattern = paste0(x, ":"))]
                 rd_chr <- rowRanges(vcf_chrom)
                 tar_chr <- as.vector(GenomicRanges::seqnames(rd_chr)@values)
                 if (length(tar_chr) > 0) {
                   tar_chr <- gsub("chr", "", tar_chr)
                   my_snps <- snpsBySeqname(all_snps, c(tar_chr))
                   snp_ID <- data.frame(posIDX = paste0("chr",
                                                        GenomicRanges::seqnames(my_snps),
                                                        ":",
                                                        pos(my_snps)),
                                        rsID = my_snps$RefSNP_id,
                                        stringsAsFactors = FALSE)
                   matV1 <- data.frame(Variant = names(rd_chr), stringsAsFactors = FALSE)
                   matV1$chromosome <- gsub("(.*):(.*)_(.*)/(.*)", "\\1", matV1$Variant)
                   matV1$start <- gsub("(.*):(.*)_(.*)/(.*)", "\\2", matV1$Variant)
                   matV1$end <- gsub("(.*):(.*)_(.*)/(.*)", "\\2", matV1$Variant)
                   matV1$ref_allele <- gsub("(.*):(.*)_(.*)/(.*)", "\\3", matV1$Variant)
                   matV1$alt_allele <- gsub("(.*):(.*)_(.*)/(.*)", "\\4", matV1$Variant)
                   matV1$posIDX <- gsub("(.*)_(.*)", "\\1", matV1$Variant)
                   matV1$start <- as.integer(matV1$start)
                   length <- 1:length(matV1$start)
                   for (i in length) {
                     matV1$end[i] <-
                       matV1$start[i] + (nchar(matV1$ref_allele[i]) - nchar(matV1$alt_allele[i]))
                   }
                   matV1$end[matV1$end < matV1$start] <- matV1$start[matV1$end < matV1$start] - 1
                   matV1$posIDX <- gsub("(.*)_(.*)", "\\1", paste0(matV1$chromosome,
                                                                   ":",
                                                                   matV1$start,
                                                                   "-",
                                                                   matV1$end))
                   matS <- merge(matV1,
                                 snp_ID[!duplicated(snp_ID[, "posIDX"]),],
                                 all.x=TRUE,
                                 by="posIDX")
                   matS <- dplyr::select(matS,-posIDX)
                   vcfID <- merge(vcfID, matS, all.x = TRUE, all.y = TRUE)
                   vcfID <- merge(vcfID, matS, all.x = TRUE, all.y = TRUE)
                 }
               }
               print("Injecting rsIDs into VCF...")
               vcfID$chr <-  as.integer(gsub("chr", "", vcfID$chromosome))
               vcfID$chr[vcfID$chromosome == "chrX"] <- 23L
               vcfID$chr[vcfID$chromosome == "chrY"] <- 24L
               vcfID$chr[vcfID$chromosome == "chrM"] <- 25L
               vcfID$chr[vcfID$chromosome == "chrMT"] <- 26L
               dd <- vcfID[order(vcfID$chr, as.integer(vcfID$start)),]
               dd$rsID[is.na(dd$rsID)] <- "."
               rsID <- dd$rsID[dd$chromosome != "chrMT"]
               if (length(vcfID$chr[vcfID$chromosome == "chrMT"]) > 0) {
                 M <- rep(".", length(names(vcf[GenomicRanges::seqnames(vcf) == "chrM",])))
                 rsID <- append(rsID, M)
                 MT <- dd$rsID[dd$chromosome == "chrMT"]
                 rsID <- append(rsID, MT)
               }
               fill <- rep(".", length(names(vcf)) - length(rsID))
               rsID <- append(rsID, fill)
               names(vcf) <- rsID

               print("Writing to VCF file...")
               setwd(outputdir)
               VariantAnnotation::writeVcf(vcf, paste0(outputdir, "/",
                                    basename(tools::file_path_sans_ext(i)),
                                    "_INDEL", ".vcf"),
                        index = TRUE)
               file.remove(list.files(normalizePath(tempfolder), full.names = TRUE))
               print(paste0("Created ", tools::file_path_sans_ext(basename(i)), "_INDEL", ".vcf.bgz"))
               file.remove(normalizePath(tempfolder))
               gc()
               setwd(wd)
               append(output, paste0(outputdir, "/",
                                     tools::file_path_sans_ext(basename(i)),
                                     "_INDEL", ".vcf.bgz"))

             }

           },
           "2"={
             ##Homo sapiens hg38
             #library(BSgenome.Hsapiens.UCSC.hg38)
             #library(SNPlocs.Hsapiens.dbSNP151.GRCh38)
             if (!requireNamespace("BSgenome.Hsapiens.UCSC.hg38", quietly = TRUE)) {
               stop("The package 'BSgenome.Hsapiens.UCSC.hg38' is required but not installed.
       Install it with: BiocManager::install('BSgenome.Hsapiens.UCSC.hg38')")
             }
             if (!requireNamespace("XtraSNPlocs.Hsapiens.dbSNP144.GRCh38", quietly = TRUE)) {
               stop("The package 'XtraSNPlocs.Hsapiens.dbSNP144.GRCh38' is required but not installed.
       Install it with: BiocManager::install('XtraSNPlocs.Hsapiens.dbSNP144.GRCh38')")
             }
             organism <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38

             ##Selects hg38 as the reference genome
             ##If reference doesn't exist within package directory, create one
             if (dir.exists(paste0(find.package("exvar"), "/hg38"))) {
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"))
             } else {
               print("Reference genome not found. Creating reference. This might take a while...")
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"),
                                    create = TRUE)
             }
             output <- c()
             for(i in bam) {
               print(paste0("Analysing ", basename(i)))
               if (isFALSE(file.exists(paste0(i, ".bai")))) {
                 print("Indexing bam...")
                 setwd(dirname(i))
                 Rsamtools::sortBam(i, tools::file_path_sans_ext(i))
                 Rsamtools::indexBam(i)
               }
               tempfolder <- paste0(dirname(i), "/temp")
               dir.create(tempfolder)
               vcflist <- c()
               bpp <- BiocParallel::MulticoreParam(threads)
               chromosomes <- GenomeInfoDb::seqlevels(refgen)
               bamfl <- Rsamtools::BamFile(i)
               tally.Param <- VariantTools::TallyVariantsParam(refgen, high_base_quality = 23L, indels = TRUE)
               print("Tallying BAM file...")
               tallies <- VariantTools::tallyVariants(bamfl, tally.Param, BPPARAM = bpp)
               gc()
               calling.filters <- VariantTools::VariantCallingFilters()
               post.filters <- VariantTools::VariantPostFilters()
               print("Calling variants...")
               snp <- VariantTools::callVariants(tallies, calling.filters, post.filters)
               sampleNames(snp) <- tools::file_path_sans_ext(dirname(i))
               mcols(snp) <- NULL
               print("Formatting variant information as VCF...")
               vcf <- VariantAnnotation::asVCF(snp)
               VariantAnnotation::writeVcf(vcf, paste0(tempfolder, "/",
                                    basename(tools::file_path_sans_ext(i)),
                                    "_INDEL", ".vcf"))


               vcflist <- c()
               vcf <- VariantAnnotation::readVcf(paste0(tempfolder, "/",
                                     basename(tools::file_path_sans_ext(i)),
                                     "_INDEL", ".vcf"))

               ##Due to constraints in memory, rsIDs are obtained on a chromosome by chromosome
               ##basis
               ##The resulting data frame is sorted as per the vcf and the rsIDs are injected
               ##into the vcf before writing it as a file again
               print("Loading dbSNP information...")
               all_snps <- XtraSNPlocs.Hsapiens.dbSNP144.GRCh38
               vcfID <- data.frame()
               mainChromosomes <- paste0("chr", GenomeInfoDb::seqlevels(all_snps))

               print("Finding rsIDs...")
               for (x in mainChromosomes) {
                 vcf_chrom <- vcf[grepl(names(vcf),
                                        pattern = paste0(x, ":"))]
                 rd_chr <- rowRanges(vcf_chrom)
                 tar_chr <- as.vector(GenomicRanges::seqnames(rd_chr)@values)
                 if (length(tar_chr) > 0) {
                   tar_chr <- gsub("chr", "", tar_chr)
                   my_snps <- snpsBySeqname(all_snps, c(tar_chr))
                   snp_ID <- data.frame(posIDX = paste0("chr",
                                                        GenomicRanges::seqnames(my_snps),
                                                        ":",
                                                        pos(my_snps)),
                                        rsID = my_snps$RefSNP_id,
                                        stringsAsFactors = FALSE)
                   matV1 <- data.frame(Variant = names(rd_chr), stringsAsFactors = FALSE)
                   matV1$chromosome <- gsub("(.*):(.*)_(.*)/(.*)", "\\1", matV1$Variant)
                   matV1$start <- gsub("(.*):(.*)_(.*)/(.*)", "\\2", matV1$Variant)
                   matV1$end <- gsub("(.*):(.*)_(.*)/(.*)", "\\2", matV1$Variant)
                   matV1$ref_allele <- gsub("(.*):(.*)_(.*)/(.*)", "\\3", matV1$Variant)
                   matV1$alt_allele <- gsub("(.*):(.*)_(.*)/(.*)", "\\4", matV1$Variant)
                   matV1$posIDX <- gsub("(.*)_(.*)", "\\1", matV1$Variant)
                   matV1$start <- as.integer(matV1$start)
                   length <- 1:length(matV1$start)
                   for (i in length) {
                     matV1$end[i] <-
                       matV1$start[i] + (nchar(matV1$ref_allele[i]) - nchar(matV1$alt_allele[i]))
                   }
                   matV1$end[matV1$end < matV1$start] <- matV1$start[matV1$end < matV1$start] - 1
                   matV1$posIDX <- gsub("(.*)_(.*)", "\\1", paste0(matV1$chromosome,
                                                                   ":",
                                                                   matV1$start,
                                                                   "-",
                                                                   matV1$end))
                   matS <- merge(matV1,
                                 snp_ID[!duplicated(snp_ID[, "posIDX"]),],
                                 all.x=TRUE,
                                 by="posIDX")
                   matS <- dplyr::select(matS,-posIDX)
                   vcfID <- merge(vcfID, matS, all.x = TRUE, all.y = TRUE)
                   vcfID <- merge(vcfID, matS, all.x = TRUE, all.y = TRUE)
                 }
               }
               print("Injecting rsIDs into VCF...")
               vcfID$chr <-  as.integer(gsub("chr", "", vcfID$chromosome))
               vcfID$chr[vcfID$chromosome == "chrX"] <- 23L
               vcfID$chr[vcfID$chromosome == "chrY"] <- 24L
               vcfID$chr[vcfID$chromosome == "chrM"] <- 25L
               vcfID$chr[vcfID$chromosome == "chrMT"] <- 26L
               dd <- vcfID[order(vcfID$chr, as.integer(vcfID$start)),]
               dd$rsID[is.na(dd$rsID)] <- "."
               rsID <- dd$rsID[dd$chromosome != "chrMT"]
               if (length(vcfID$chr[vcfID$chromosome == "chrMT"]) > 0) {
                 M <- rep(".", length(names(vcf[GenomicRanges::seqnames(vcf) == "chrM",])))
                 rsID <- append(rsID, M)
                 MT <- dd$rsID[dd$chromosome == "chrMT"]
                 rsID <- append(rsID, MT)
               }
               fill <- rep(".", length(names(vcf)) - length(rsID))
               rsID <- append(rsID, fill)
               names(vcf) <- rsID

               print("Writing to VCF file...")
               setwd(outputdir)
               VariantAnnotation::writeVcf(vcf, paste0(outputdir, "/",
                                    basename(tools::file_path_sans_ext(i)),
                                    "_INDEL", ".vcf"),
                        index = TRUE)
               file.remove(list.files(normalizePath(tempfolder), full.names = TRUE))
               print(paste0("Created ", tools::file_path_sans_ext(basename(i)), "_INDEL", ".vcf.bgz"))
               file.remove(normalizePath(tempfolder))
               gc()
               setwd(wd)
               append(output, paste0(outputdir, "/",
                                     tools::file_path_sans_ext(basename(i)),
                                     "_INDEL", ".vcf.bgz"))

             }
           },
           "3"={
             ##Mus musculus mm10
             #library(BSgenome.Mmusculus.UCSC.mm10)
             if (!requireNamespace("BSgenome.Mmusculus.UCSC.mm10", quietly = TRUE)) {
               stop("The package 'BSgenome.Mmusculus.UCSC.mm10' is required but not installed.
       Install it with: BiocManager::install('BSgenome.Mmusculus.UCSC.mm10')")
             }

             organism <- BSgenome.Mmusculus.UCSC.mm10::BSgenome.Mmusculus.UCSC.mm10

             ##Selects mm10 as the reference genome
             ##If reference doesn't exist within package directory, create one
             if (dir.exists(paste0(find.package("exvar"), "/mm10"))) {
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"))
             } else {
               print("Reference genome not found. Creating reference. This might take a while...")
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"),
                                    create = TRUE)
             }
             output <- c()
             output <- c()
             for(i in bam) {
               print(paste0("Analysing ", basename(i)))
               if (isFALSE(file.exists(paste0(i, ".bai")))) {
                 print("Indexing bam...")
                 setwd(dirname(i))
                 Rsamtools::sortBam(i, tools::file_path_sans_ext(i))
                 Rsamtools::indexBam(i)
               }
               tempfolder <- paste0(dirname(i), "/temp")
               dir.create(tempfolder)
               vcflist <- c()
               bpp <- BiocParallel::MulticoreParam(threads)
               chromosomes <- GenomeInfoDb::seqlevels(refgen)
               bamfl <- Rsamtools::BamFile(i)
               tally.Param <- VariantTools::TallyVariantsParam(refgen, high_base_quality = 23L, indels = TRUE)
               print("Tallying BAM file...")
               tallies <- VariantTools::tallyVariants(bamfl, tally.Param, BPPARAM = bpp)
               gc()
               calling.filters <- VariantTools::VariantCallingFilters()
               post.filters <- VariantTools::VariantPostFilters()
               print("Calling variants...")
               snp <- VariantTools::callVariants(tallies, calling.filters, post.filters)
               sampleNames(snp) <- tools::file_path_sans_ext(dirname(i))
               mcols(snp) <- NULL
               print("Formatting variant information as VCF...")
               vcf <- VariantAnnotation::asVCF(snp)
               print("Writing to VCF file...")
               setwd(outputdir)
               VariantAnnotation::writeVcf(vcf, paste0(outputdir, "/",
                                    basename(tools::file_path_sans_ext(i)),
                                    "_INDEL", ".vcf"),
                        index = TRUE)
               file.remove(list.files(normalizePath(tempfolder), full.names = TRUE))
               print(paste0("Created ", tools::file_path_sans_ext(basename(i)), "_INDEL", ".vcf.bgz"))
               file.remove(normalizePath(tempfolder))
               gc()
               setwd(wd)
               append(output, paste0(outputdir, "/",
                                     tools::file_path_sans_ext(basename(i)),
                                     "_INDEL", ".vcf.bgz"))
             }
           },
           "4"={
             ##Arabidopsis thaliana TAIR9
            # library(BSgenome.Athaliana.TAIR.TAIR9)
             if (!requireNamespace("BSgenome.Athaliana.TAIR.TAIR9", quietly = TRUE)) {
               stop("The package 'BSgenome.Athaliana.TAIR.TAIR9' is required but not installed.
       Install it with: BiocManager::install('BSgenome.Athaliana.TAIR.TAIR9')")
             }

             organism <- BSgenome.Athaliana.TAIR.TAIR9::BSgenome.Athaliana.TAIR.TAIR9

             ##Selects hg19 as the reference genome
             ##If reference doesn't exist within package directory, create one
             if (dir.exists(paste0(find.package("exvar"), "/TAIR9"))) {
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"))
             } else {
               print("Reference genome not found. Creating reference. This might take a while...")
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"),
                                    create = TRUE)
             }
             output <- c()
             output <- c()
             for(i in bam) {
               print(paste0("Analysing ", basename(i)))
               if (isFALSE(file.exists(paste0(i, ".bai")))) {
                 print("Indexing bam...")
                 setwd(dirname(i))
                 Rsamtools::sortBam(i, tools::file_path_sans_ext(i))
                 Rsamtools::indexBam(i)
               }
               tempfolder <- paste0(dirname(i), "/temp")
               dir.create(tempfolder)
               vcflist <- c()
               bpp <- BiocParallel::MulticoreParam(threads)
               chromosomes <- GenomeInfoDb::seqlevels(refgen)
               bamfl <- Rsamtools::BamFile(i)
               tally.Param <- VariantTools::TallyVariantsParam(refgen, high_base_quality = 23L, indels = TRUE)
               print("Tallying BAM file...")
               tallies <- VariantTools::tallyVariants(bamfl, tally.Param, BPPARAM = bpp)
               gc()
               calling.filters <- VariantTools::VariantCallingFilters()
               post.filters <- VariantTools::VariantPostFilters()
               print("Calling variants...")
               snp <- VariantTools::callVariants(tallies, calling.filters, post.filters)
               sampleNames(snp) <- tools::file_path_sans_ext(dirname(i))
               mcols(snp) <- NULL
               print("Formatting variant information as VCF...")
               vcf <- VariantAnnotation::asVCF(snp)
               print("Writing to VCF file...")
               setwd(outputdir)
               VariantAnnotation::writeVcf(vcf, paste0(outputdir, "/",
                                    basename(tools::file_path_sans_ext(i)),
                                    "_INDEL", ".vcf"),
                        index = TRUE)
               file.remove(list.files(normalizePath(tempfolder), full.names = TRUE))
               print(paste0("Created ", tools::file_path_sans_ext(basename(i)), "_INDEL", ".vcf.bgz"))
               file.remove(normalizePath(tempfolder))
               gc()
               setwd(wd)
               append(output, paste0(outputdir, "/",
                                     tools::file_path_sans_ext(basename(i)),
                                     "_INDEL", ".vcf.bgz"))
             }
           },
           "5"={
             ##Drosophilia melanogaster dm6
             #library(BSgenome.Dmelanogaster.UCSC.dm6)
             if (!requireNamespace("BSgenome.Dmelanogaster.UCSC.dm6", quietly = TRUE)) {
               stop("The package 'BSgenome.Dmelanogaster.UCSC.dm6' is required but not installed.
       Install it with: BiocManager::install('BSgenome.Dmelanogaster.UCSC.dm6')")
             }

             organism <- BSgenome.Dmelanogaster.UCSC.dm6::BSgenome.Dmelanogaster.UCSC.dm6

             ##Selects dm6 as the reference genome
             ##If reference doesn't exist within package directory, create one
             if (dir.exists(paste0(find.package("exvar"), "/dm6"))) {
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"))
             } else {
               print("Reference genome not found. Creating reference. This might take a while...")
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"),
                                    create = TRUE)
             }
             output <- c()
             output <- c()
             for(i in bam) {
               print(paste0("Analysing ", basename(i)))
               if (isFALSE(file.exists(paste0(i, ".bai")))) {
                 print("Indexing bam...")
                 setwd(dirname(i))
                 Rsamtools::sortBam(i, tools::file_path_sans_ext(i))
                 Rsamtools::indexBam(i)
               }
               tempfolder <- paste0(dirname(i), "/temp")
               dir.create(tempfolder)
               vcflist <- c()
               bpp <- BiocParallel::MulticoreParam(threads)
               chromosomes <- GenomeInfoDb::seqlevels(refgen)
               bamfl <- Rsamtools::BamFile(i)
               tally.Param <- VariantTools::TallyVariantsParam(refgen, high_base_quality = 23L, indels = TRUE)
               print("Tallying BAM file...")
               tallies <- VariantTools::tallyVariants(bamfl, tally.Param, BPPARAM = bpp)
               gc()
               calling.filters <- VariantTools::VariantCallingFilters()
               post.filters <- VariantTools::VariantPostFilters()
               print("Calling variants...")
               snp <- VariantTools::callVariants(tallies, calling.filters, post.filters)
               sampleNames(snp) <- tools::file_path_sans_ext(dirname(i))
               mcols(snp) <- NULL
               print("Formatting variant information as VCF...")
               vcf <- VariantAnnotation::asVCF(snp)
               print("Writing to VCF file...")
               setwd(outputdir)
               VariantAnnotation::writeVcf(vcf, paste0(outputdir, "/",
                                    basename(tools::file_path_sans_ext(i)),
                                    "_INDEL", ".vcf"),
                        index = TRUE)
               file.remove(list.files(normalizePath(tempfolder), full.names = TRUE))
               print(paste0("Created ", tools::file_path_sans_ext(basename(i)), "_INDEL", ".vcf.bgz"))
               file.remove(normalizePath(tempfolder))
               gc()
               setwd(wd)
               append(output, paste0(outputdir, "/",
                                     tools::file_path_sans_ext(basename(i)),
                                     "_INDEL", ".vcf.bgz"))
             }
           },
           "6"={
             ##Danio rerio danRer11
            # library(BSgenome.Drerio.UCSC.danRer11)
             if (!requireNamespace("BSgenome.Drerio.UCSC.danRer11", quietly = TRUE)) {
               stop("The package 'BSgenome.Drerio.UCSC.danRer11' is required but not installed.
       Install it with: BiocManager::install('BSgenome.Drerio.UCSC.danRer11')")
             }

             organism <- BSgenome.Drerio.UCSC.danRer11::BSgenome.Drerio.UCSC.danRer11

             ##Selects danRer11 as the reference genome
             ##If reference doesn't exist within package directory, create one
             if (dir.exists(paste0(find.package("exvar"), "/danRer11"))) {
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"))
             } else {
               print("Reference genome not found. Creating reference. This might take a while...")
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"),
                                    create = TRUE)
             }
             output <- c()
             for(i in bam) {
               print(paste0("Analysing ", basename(i)))
               if (isFALSE(file.exists(paste0(i, ".bai")))) {
                 print("Indexing bam...")
                 setwd(dirname(i))
                 Rsamtools::sortBam(i, tools::file_path_sans_ext(i))
                 Rsamtools::indexBam(i)
               }
               tempfolder <- paste0(dirname(i), "/temp")
               dir.create(tempfolder)
               vcflist <- c()
               bpp <- BiocParallel::MulticoreParam(threads)
               chromosomes <- GenomeInfoDb::seqlevels(refgen)
               bamfl <- Rsamtools::BamFile(i)
               tally.Param <- VariantTools::TallyVariantsParam(refgen, high_base_quality = 23L, indels = TRUE)
               print("Tallying BAM file...")
               tallies <- VariantTools::tallyVariants(bamfl, tally.Param, BPPARAM = bpp)
               gc()
               calling.filters <- VariantTools::VariantCallingFilters()
               post.filters <- VariantTools::VariantPostFilters()
               print("Calling variants...")
               snp <- VariantTools::callVariants(tallies, calling.filters, post.filters)
               sampleNames(snp) <- tools::file_path_sans_ext(dirname(i))
               mcols(snp) <- NULL
               print("Formatting variant information as VCF...")
               vcf <- VariantAnnotation::asVCF(snp)
               print("Writing to VCF file...")
               setwd(outputdir)
               VariantAnnotation::writeVcf(vcf, paste0(outputdir, "/",
                                    basename(tools::file_path_sans_ext(i)),
                                    "_INDEL", ".vcf"),
                        index = TRUE)
               file.remove(list.files(normalizePath(tempfolder), full.names = TRUE))
               print(paste0("Created ", tools::file_path_sans_ext(basename(i)), "_INDEL", ".vcf.bgz"))
               file.remove(normalizePath(tempfolder))
               gc()
               setwd(wd)
               append(output, paste0(outputdir, "/",
                                     tools::file_path_sans_ext(basename(i)),
                                     "_INDEL", ".vcf.bgz"))
             }
           },
           "7"={
             ##Rattus norvegicus rn5
             #library(BSgenome.Rnorvegicus.UCSC.rn5)
             if (!requireNamespace("BSgenome.Rnorvegicus.UCSC.rn5", quietly = TRUE)) {
               stop("The package 'BSgenome.Rnorvegicus.UCSC.rn5' is required but not installed.
       Install it with: BiocManager::install('BSgenome.Rnorvegicus.UCSC.rn5')")
             }

             organism <- BSgenome.Rnorvegicus.UCSC.rn5::BSgenome.Rnorvegicus.UCSC.rn5

             ##Selects danRer11 as the reference genome
             ##If reference doesn't exist within package directory, create one
             if (dir.exists(paste0(find.package("exvar"), "/rn5"))) {
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"))
             } else {
               print("Reference genome not found. Creating reference. This might take a while...")
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"),
                                    create = TRUE)
             }
             output <- c()
             for(i in bam) {
               print(paste0("Analysing ", basename(i)))
               if (isFALSE(file.exists(paste0(i, ".bai")))) {
                 print("Indexing bam...")
                 setwd(dirname(i))
                 Rsamtools::sortBam(i, tools::file_path_sans_ext(i))
                 Rsamtools::indexBam(i)
               }
               tempfolder <- paste0(dirname(i), "/temp")
               dir.create(tempfolder)
               vcflist <- c()
               bpp <- BiocParallel::MulticoreParam(threads)
               chromosomes <- GenomeInfoDb::seqlevels(refgen)
               bamfl <- Rsamtools::BamFile(i)
               tally.Param <- VariantTools::TallyVariantsParam(refgen, high_base_quality = 23L, indels = TRUE)
               print("Tallying BAM file...")
               tallies <- VariantTools::tallyVariants(bamfl, tally.Param, BPPARAM = bpp)
               gc()
               calling.filters <- VariantTools::VariantCallingFilters()
               post.filters <- VariantTools::VariantPostFilters()
               print("Calling variants...")
               snp <- VariantTools::callVariants(tallies, calling.filters, post.filters)
               sampleNames(snp) <- tools::file_path_sans_ext(dirname(i))
               mcols(snp) <- NULL
               print("Formatting variant information as VCF...")
               vcf <- VariantAnnotation::asVCF(snp)
               print("Writing to VCF file...")
               setwd(outputdir)
               VariantAnnotation::writeVcf(vcf, paste0(outputdir, "/",
                                    basename(tools::file_path_sans_ext(i)),
                                    "_INDEL", ".vcf"),
                        index = TRUE)
               file.remove(list.files(normalizePath(tempfolder), full.names = TRUE))
               print(paste0("Created ", tools::file_path_sans_ext(basename(i)), "_INDEL", ".vcf.bgz"))
               file.remove(normalizePath(tempfolder))
               gc()
               setwd(wd)
               append(output, paste0(outputdir, "/",
                                     tools::file_path_sans_ext(basename(i)),
                                     "_INDEL", ".vcf.bgz"))
             }
           },
           "8"={
             ##Saccharomyces cerevisiae sacCer3
             #library(BSgenome.Scerevisiae.UCSC.sacCer3)
             if (!requireNamespace("BSgenome.Scerevisiae.UCSC.sacCer3", quietly = TRUE)) {
               stop("The package 'BSgenome.Scerevisiae.UCSC.sacCer3' is required but not installed.
       Install it with: BiocManager::install('BSgenome.Scerevisiae.UCSC.sacCer3')")
             }

             organism <- BSgenome.Scerevisiae.UCSC.sacCer3::BSgenome.Scerevisiae.UCSC.sacCer3

             ##Selects sacCer3 as the reference genome
             ##If reference doesn't exist within package directory, create one
             if (dir.exists(paste0(find.package("exvar"), "/sacCer3"))) {
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"))
             } else {
               print("Reference genome not found. Creating reference. This might take a while...")
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"),
                                    create = TRUE)
             }
             ouput <- c()
             output <- c()
             for(i in bam) {
               print(paste0("Analysing ", basename(i)))
               if (isFALSE(file.exists(paste0(i, ".bai")))) {
                 print("Indexing bam...")
                 setwd(dirname(i))
                 Rsamtools::sortBam(i, tools::file_path_sans_ext(i))
                 Rsamtools::indexBam(i)
               }
               tempfolder <- paste0(dirname(i), "/temp")
               dir.create(tempfolder)
               vcflist <- c()
               bpp <- BiocParallel::MulticoreParam(threads)
               chromosomes <- GenomeInfoDb::seqlevels(refgen)
               bamfl <- Rsamtools::BamFile(i)
               tally.Param <- VariantTools::TallyVariantsParam(refgen, high_base_quality = 23L, indels = TRUE)
               print("Tallying BAM file...")
               tallies <- VariantTools::tallyVariants(bamfl, tally.Param, BPPARAM = bpp)
               gc()
               calling.filters <- VariantTools::VariantCallingFilters()
               post.filters <- VariantTools::VariantPostFilters()
               print("Calling variants...")
               snp <- VariantTools::callVariants(tallies, calling.filters, post.filters)
               sampleNames(snp) <- tools::file_path_sans_ext(dirname(i))
               mcols(snp) <- NULL
               print("Formatting variant information as VCF...")
               vcf <- VariantAnnotation::asVCF(snp)
               print("Writing to VCF file...")
               setwd(outputdir)
               VariantAnnotation::writeVcf(vcf, paste0(outputdir, "/",
                                    basename(tools::file_path_sans_ext(i)),
                                    "_INDEL", ".vcf"),
                        index = TRUE)
               file.remove(list.files(normalizePath(tempfolder), full.names = TRUE))
               print(paste0("Created ", tools::file_path_sans_ext(basename(i)), "_INDEL", ".vcf.bgz"))
               file.remove(normalizePath(tempfolder))
               gc()
               setwd(wd)
               append(output, paste0(outputdir, "/",
                                     tools::file_path_sans_ext(basename(i)),
                                     "_INDEL", ".vcf.bgz"))
             }
           },
           "9"={
             ##Caenorhabditis elagans
            # library(BSgenome.Celegans.UCSC.ce11)
             if (!requireNamespace("BSgenome.Celegans.UCSC.ce11", quietly = TRUE)) {
               stop("The package 'BSgenome.Celegans.UCSC.ce11' is required but not installed.
       Install it with: BiocManager::install('BSgenome.Celegans.UCSC.ce11')")
             }

             organism <- BSgenome.Celegans.UCSC.ce11::BSgenome.Celegans.UCSC.ce11

             ##Selects ce11 as the reference genome
             ##If reference doesn't exist within package directory, create one
             if (dir.exists(paste0(find.package("exvar"), "/ce11"))) {
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"))
             } else {
               print("Reference genome not found. Creating reference. This might take a while...")
               refgen <- gmapR::GmapGenome(organism,
                                    directory = find.package("exvar"),
                                    create = TRUE)
             }
             output <- c()
             for(i in bam) {
               print(paste0("Analysing ", basename(i)))
               if (isFALSE(file.exists(paste0(i, ".bai")))) {
                 print("Indexing bam...")
                 setwd(dirname(i))
                 Rsamtools::sortBam(i, tools::file_path_sans_ext(i))
                 Rsamtools::indexBam(i)
               }
               tempfolder <- paste0(dirname(i), "/temp")
               dir.create(tempfolder)
               vcflist <- c()
               bpp <- BiocParallel::MulticoreParam(threads)
               chromosomes <- GenomeInfoDb::seqlevels(refgen)
               bamfl <- Rsamtools::BamFile(i)
               tally.Param <- VariantTools::TallyVariantsParam(refgen, high_base_quality = 23L, indels = TRUE)
               print("Tallying BAM file...")
               tallies <- VariantTools::tallyVariants(bamfl, tally.Param, BPPARAM = bpp)
               gc()
               calling.filters <- VariantTools::VariantCallingFilters()
               post.filters <- VariantTools::VariantPostFilters()
               print("Calling variants...")
               snp <- VariantTools::callVariants(tallies, calling.filters, post.filters)
               sampleNames(snp) <- tools::file_path_sans_ext(dirname(i))
               mcols(snp) <- NULL
               print("Formatting variant information as VCF...")
               vcf <- VariantAnnotation::asVCF(snp)
               print("Writing to VCF file...")
               setwd(outputdir)
               VariantAnnotation::writeVcf(vcf, paste0(outputdir, "/",
                                    basename(tools::file_path_sans_ext(i)),
                                    "_INDEL", ".vcf"),
                        index = TRUE)
               file.remove(list.files(normalizePath(tempfolder), full.names = TRUE))
               print(paste0("Created ", tools::file_path_sans_ext(basename(i)), "_INDEL", ".vcf.bgz"))
               file.remove(normalizePath(tempfolder))
               gc()
               setwd(wd)
               append(output, paste0(outputdir, "/",
                                     tools::file_path_sans_ext(basename(i)),
                                     "_INDEL", ".vcf.bgz"))
             }
           }
    )
  }
  setwd(wd)
  return(output)
}


